doc/tutorial.md

```diff
@@ -0,0 +1,594 @@
+* [SNAT Mode (two-arm)](#snat)
+* [Bonding and VLAN devices](#bond-vlan)
+
+> To compile and launch DPVS, pls check *README.md* for this project.
+
+<a id='term'/>
+
+# Terminology
+
+For the meanings of *Full-NAT* (`FNAT`), `DR`, `toa`, `OSPF`/`ECMP` and `keepalived`, pls refer [LVS](www.linuxvirtualserver.org) and [Alibaba/LVS](https://github.com/alibaba/LVS/tree/master/docs).
+
+Note `DPVS` support `FNAT`, `DR`, `SNAT` forwarding modes, and each mode can be configured as `one-arm` or `two-arm` topology, with or without `ospfd`/`keepalived`. There're too many combinations, I cannot list all the examples here. Let's just give some popular working models used in our daily work.
+
+> `Tunneling` and `SNAT` is not supported because we didn't see strong demand yet, at least from *iQiYi*. May add them in the future. :)
+
+<a id='one-two-arm'/>
+
+## *One-arm* and *two-arm*
+
+The term *two-arm* means, you have clients in one side of *load-balancer* (`LB`) and servers (`RS`) in another side, then `LB` forwards packets between its two logical network interfaces. For example, *WAN-to-LAN* load balancing.
+
+On the other hand, *one-arm* means all clients and servers are in same side of `load-balancer`, `LB` forwards traffic through the same logical network interface.
+
+> *Logical interface* (or *device*) could be physical `DPDK` interface, or `DPVS` virtual devices like *Bonding* and *VLAN* devices.
+
+To make things easier, we do not consider virtual devices for now. Thus, *two-arm* topology need
+
+* two DPDK interfaces loaded with `igb_uio` driver, and
+* `/etc/dpvs.conf` should also be configured with two interfaces. Pls refer the file `conf/dpvs.conf.sample`.
+
+```
+$ dpdk-devbind --st
+
+Network devices using DPDK-compatible driver
+============================================
+0000:06:00.0 'Ethernet Controller 10-Gigabit X540-AT2' drv=igb_uio unused=uio_pci_generic
+0000:06:00.1 'Ethernet Controller 10-Gigabit X540-AT2' drv=igb_uio unused=uio_pci_generic
+```
+
+For *one-arm*, only one DPDK intreface needed, and you can refer `conf/dpvs.conf.single-nic.sample`.
+
+<a id='kni'/>
+
+## KNI Device
+
+Like `LVS`, `DPVS` can be deployed as different sort of *Cluster* models for High-Available (HA) purpose. Both *OSPF/ECMP* and *Master/Backup* models are supported. *OSPF/ECMP* model need package `quagga` and its `zebra` and `ospfd` programs. And *master/back* model need `Keepalived`.
+
+Consider `DPDK` application manages the networking interface completely (except the extra control NIC if exist), Linux Kernel and programs run on Kernel TCP/IP stack cannot receive packets from `DPDK` interface directly. To make Linux programs like `sshd`, `zebra/ospfd` and `keepalived` work, DPDK `kni` device is used. Then the Linux programs can working on `kni` device with Linux TCP/IP stack. Actually, `DPVS` pass the packets it's not interested in to `kni` device, like `OSPF/VRRP/ssh` packets. So that the programs "listening" on Linux stack are able to handle them.
```

Comment:
Considering ...

Reply:
`DPVS` pass the packets to `kni` device  which it's not interested in, like `OSPF/VRRP/ssh` packets. So the programs ...